ska_trend/centroid_dashboard/app.py

```diff
@@ -0,0 +1,1038 @@
+        entry["dzag_median"] = np.median(crs[slot].dzags)
+        mags = fetch.Msid(f"aoacmag{slot}", start, stop)
+        entry["mag_median"] = np.median(mags.vals)
+
+
+def write_info_json(obs: Observation, info_json_path: Path):
+    """Write the info.json file for the observation."""
+    out = obs.info.copy()
+    for key, val in out.items():
+        if isinstance(val, float):
+            out[key] = round(val, 3)
+    logger.info(f"Writing {info_json_path}")
+    info_json_path.write_text(json.dumps(out, indent=2, sort_keys=True))
+
+
+def write_centroid_resids(crs: dict[int, CentroidResiduals], save_path: Path) -> None:
+    """Write the centroid residuals to a file.
+
+    This creates a dictionary with the slot number as the key and a dictionary with the
+    start and stop times, the number of times, and the dyags and dzags as the values.
+    The structure is::
+
+    {
+        slot: {
+            "tstart": t0,
+            "tstop": t1,
+            "n_times": n_times,
+            "dyags": dyags,
+            "dzags": dzags,
+        },
+    }
+
+    Notes:
+    - The dyags and dzags are interpolated to a common time grid based on the median
+      difference between the yag times.
+    - The times are in CXC seconds.
+    - The dyags and dzags are in arcseconds.
+    - The output is a pickle file that can be loaded later.
+    - The file is saved in the report directory for the observation.
+    - The dyags and dzags are saved as float16 to save space. The max error is around
+      0.002 arcsec.
+
+    Parameters
+    ----------
+    crs : dict
+        Dictionary of CentroidResiduals objects keyed by slot.
+    save_path : Path
+        Path to save the centroid residuals file.
+    """
+    out = {}
+    for slot, cr in crs.items():
+        t0 = max(cr.yag_times[0], cr.zag_times[0])
+        t1 = min(cr.yag_times[-1], cr.zag_times[-1])
+        ok = (cr.yag_times >= t0) & (cr.yag_times <= t1)
+        if np.count_nonzero(ok) < 4:
+            logger.warning(
+                f"Not enough points in slot {slot} to write centroid residuals"
+            )
+            continue
+        dt_median = np.median(np.diff(cr.yag_times[ok]))
+        n_times = int(np.round(t1 - t0) / dt_median) + 1
+        times = np.linspace(t0, t1, n_times)
+        info_slot = {"tstart": t0, "tstop": t1, "n_times": n_times}
+        info_slot["dyags"] = np.interp(times, cr.yag_times, cr.dyags).astype(np.float16)
+        info_slot["dzags"] = np.interp(times, cr.zag_times, cr.dzags).astype(np.float16)
+        out[slot] = info_slot
+
+    logger.info(f"Writing {save_path} with {len(out)} slots")
+    save_path.write_bytes(pickle.dumps(out))
+
+
+def process_obs(obs: Observation, opt: argparse.Namespace):
+    """Process the observation."""
+    report_dir = obs.report_dir
+    report_dir.mkdir(parents=True, exist_ok=True)
+
+    if obs.manvr_event is None:
+        logger.info(f"ObsID {obs.obsid} has no maneuver event in telemetry, skipping")
```

Comment:
For other special cases in historical data and being robust, how does this behave on a undercover?

Reply:
Good question... this is probably more about how kadi observations handles this, but I'll check. It looks like JAN1121 is the last week with an undercover, at least from grepping the OR-list files.